src/main/java/com/iota/iri/network/TransactionRequester.java

```diff
@@ -84,48 +82,49 @@ private boolean transactionsToRequestIsFull() {
 
 
     public Hash transactionToRequest(boolean milestone) throws Exception {
-        final long beginningTime = System.currentTimeMillis();
+        // determine which set of transactions to operate on
+        Set<Hash> primarySet = milestone ? milestoneTransactionsToRequest : transactionsToRequest;
+        Set<Hash> alternativeSet = milestone ? transactionsToRequest : milestoneTransactionsToRequest;
+        Set<Hash> requestSet = primarySet.size() == 0 ? alternativeSet : primarySet;
+
+        // determine the first hash in our set that needs to be processed
         Hash hash = null;
-        Set<Hash> requestSet;
-        if(milestone) {
-             requestSet = milestoneTransactionsToRequest;
-             if(requestSet.size() == 0) {
-                 requestSet = transactionsToRequest;
-             }
-        } else {
-            requestSet = transactionsToRequest;
-            if(requestSet.size() == 0) {
-                requestSet = milestoneTransactionsToRequest;
-            }
-        }
         synchronized (syncObj) {
+            // repeat while we have transactions that shall be requested
             while (requestSet.size() != 0) {
+                // remove the first item in our set for further examination
                 Iterator<Hash> iterator = requestSet.iterator();
                 hash = iterator.next();
                 iterator.remove();
+
+                // if we have received the transaction in the mean time ....
                 if (TransactionViewModel.exists(tangle, hash)) {
+                    // ... dump a log message ...
                     log.info("Removed existing tx from request list: " + hash);
                     messageQ.publish("rtl %s", hash);
-                } else {
-                    if (!transactionsToRequestIsFull()) {
-                        requestSet.add(hash);
-                    }
-                    break;
+
+                    // ... and continue to the next element in the set
+                    continue;
                 }
+
+                // ... otherwise -> re-add it at the end of the set ...
+                //
+                // Note: we always have enough space since we removed the element before
```

Comment:
Are you positive about that?
In `requestTransaction` method we add without checking if we are full (which may be another bug). But if I call on this method enough times I can exceed capacity and then I won't have enough space even though I just removed a hash.

Anyhows, now you introduced a logic change that should be tested...
The reason it takes time to refactor IRI is becuase we want to add proper tests to the changes and not just replace bugs with other bugs.

I personally want to refactor every single class here but I want to do it correctly.

Reply:
I am positive about that!

In line 71 (requestTransaction) we DO check if !transactionsToRequestIsFull() if it is a normal transaction. the request set for milestoneTransactions is not limited in size, which makes sense since all of the milestones are relevant for our node.

Reply:
:+1: